### DISCUSSION TITLE
Diligent integration with Qt

### DISCUSSION THREAD

**Discussion by AnloWho** [DiligentGraphics/DiligentCore#625] [General]
I'm looking into using DiligentEngine as backend for graphics in Qt. Does anyone around here have experience in that? would it work?

**Comment by TheMostDiligent** [comment]
I don't have experience with Qt, but I think there should be no problems making it work with Diligent.
As long as you can get the window handle from the library, you can attach Diligent swap chain to it and render.

**Comment by promerosenabre** [comment]
It works fine while not opengl is not activated on your Qt App. I want to combine two qwidgets in same layout, one is QOpenGLWidget and other use diligent engine, and diligent engine doesn't work, I get error Diligent::GLContextState::BindSampler(). 
when I demote QOpenGLWidget  to QWidget, diligent engine qwidget works fine. Any idea ? I guess problem is because context is managed by qt... 
THANKS

**Reply by TheMostDiligent** [reply]
> I guess problem is because context is managed by qt

This is very likely. Try attaching to existing context created by QT with the `AttachToActiveGLContext`

https://github.com/DiligentGraphics/DiligentCore/blob/3e17a10d517ac5ca6d13e3470fb7982b7dd17bfd/Graphics/GraphicsEngineOpenGL/interface/EngineFactoryOpenGL.h#L82-L85

**Comment by promerosenabre** [comment]
Thanks, I have changed my code without sucessfull, sorry, but I think I am close to find the solution.
Anycase, I still have some questions about ( I am analizing plugun Unity on samples )

now I use 
 **pFactoryOpenGL->AttachToActiveGLContext(EngineCI, &m_pDevice, &m_pImmediateContext);**
it means I don't have to use 
**pFactoryOpenGL->CreateDeviceAndSwapChainGL(EngineCI, &m_pDevice, &m_pImmediateContext, SCDesc,&m_pSwapChain);**

if yes, what happens with m_pSwapChain how to generate it ?

I see on Unity sample :

**auto &DefaultAllocator = DefaultRawMemoryAllocator::GetAllocator();
auto pProxySwapChainGL = NEW_RC_OBJ(DefaultAllocator, "ProxySwapChainGL instance", ProxySwapChainGL)(m_UnityGraphicsGL, m_pDevice, m_pDeviceCtx, SCDesc);
pProxySwapChainGL->QueryInterface(IID_SwapChain, reinterpret_cast<IObject**>(static_cast<ISwapChain**>(&m_pProxySwapChain)));
RefCntAutoPtr<IDeviceContextGL>(m_pDeviceCtx, IID_DeviceContextGL)->SetSwapChain(pProxySwapChainGL);**

and I undestand this sample use rendertotexture and share the image between diligine and unity

Do I have to do the same because I am working with external opengl ?

or I only have to generate a correct m_pSwapChain ?

unity sample  uses m_pSwapChainGL.
and 
RefCntAutoPtr(m_pDeviceCtx, IID_DeviceContextGL)->SetSwapChain(pProxySwapChainGL);*

Sorry, I am confused about this step.

Thanks

**Reply by TheMostDiligent** [reply]
> it means I don't have to use pFactoryOpenGL->CreateDeviceAndSwapChainGL(EngineCI, &m_pDevice, &m_pImmediateContext, SCDesc,&m_pSwapChain);

Yes, you only call `AttachToActiveGLContext`

> if yes, what happens with m_pSwapChain how to generate it ?

In OpenGL, context and swap chain are both attached to the window. If your application creates the context, it is also responsible for presenting (e.g. calling `glSwapBuffers`). Diligent can't create the swap chain and present.

Unity sample may not be the best example. You application outline should look something like this:

Initialization:
* Qt initializes OpenGL context
* Call `AttachToActiveGLContext` to get Diligent device and immediate context
* Get OpenGL handle from the texture you want to share with Diligent, call [CreateTextureFromGLHandle](https://github.com/DiligentGraphics/DiligentCore/blob/3e17a10d517ac5ca6d13e3470fb7982b7dd17bfd/Graphics/GraphicsEngineOpenGL/interface/RenderDeviceGL.h#L104) to create Diligent texture

Rendering:
* Qt starts rendering
* Switch to Diligent, call [InvalidateState](https://github.com/DiligentGraphics/DiligentCore/blob/3e17a10d517ac5ca6d13e3470fb7982b7dd17bfd/Graphics/GraphicsEngine/interface/DeviceContext.h#L2540C25-L2540C40)
* Render with Diligent
* Present with Qt

**Reply by promerosenabre** [reply]
Thank you for your quick replay, I apreciate it.  
I have simplified the initialization of diligent without mistakes. Thanks

I try to replicate on Qt the sample "Tutorial03_Texturing" that works fine without opengl Qt activated.

Now I get a mistake on 
MapHelper<Constants> CBConstants(m_pImmediateContext, this->modelCube.materials[this->modelCube.items[cont].indexMaterial].m_VSConstants, MAP_WRITE, MAP_FLAG_DISCARD);

inside render.  it is sense because I use AttachToActiveGLContext ?

it works fine without opengl

any idea ? sorry for all this questions, I hope will be usefull for others than try to use diligent on Qt.

**Reply by promerosenabre** [reply]
Error solved after define correctly SetRenderTargets with apropiate ITextureView.

**Reply by AnloWho** [reply]
> Thank you for your quick replay, I apreciate it. I have simplified the initialization of diligent without mistakes. Thanks
> 
> I try to replicate on Qt the sample "Tutorial03_Texturing" that works fine without opengl Qt activated.
> 
> Now I get a mistake on MapHelper CBConstants(m_pImmediateContext, this->modelCube.materials[this->modelCube.items[cont].indexMaterial].m_VSConstants, MAP_WRITE, MAP_FLAG_DISCARD);
> 
> inside render. it is sense because I use AttachToActiveGLContext ?
> 
> it works fine without opengl
> 
> any idea ? sorry for all this questions, I hope will be usefull for others than try to use diligent on Qt.

@promerosenabre what do you mean by "without opengl" it works fine in Qt? I'm struggling with sorting out the connection between Qt and Diligent.

**Comment by AnloWho** [comment]
Very interesting. I'll also explore the land to hook up diligent to qt.
It's enough with Diligent Core if I want to do basic polygon rendering with
textures and so. Right?

**Reply by promerosenabre** [reply]
Still now working I have black screen, I undestand because I have to get the data from iTexture ( working on it ).
and draw it on Qt.

Would be great if we can see a sample about Qt integration with diligent engine. 
Polygon rendering with texture will be enought to undestand the mecanism. I am working with Qt 6.8

Jus to infom you:  if you use winId of a QWidget everything goes fine using CreateDeviceAndSwapChainGL.
it is posible to integrate it,  following Visual studio Tutorials.

But, as soon as there is a QOpenGLWidget  working on same window, not necesary the QWidget that you want to show diligent render, 
there is an external opengl context and you must use AttachToActiveGLContext, and everything becames  more complex, at least for me.

Thanks you very much for your support.

**Reply by TheMostDiligent** [reply]
Yes, OpenGL is not good when it comes to sharing the context.

> I undestand because I have to get the data from iTexture

You may use [ITextureGL::GetGLTextureHandle](https://github.com/DiligentGraphics/DiligentCore/blob/3e17a10d517ac5ca6d13e3470fb7982b7dd17bfd/Graphics/GraphicsEngineOpenGL/interface/TextureGL.h#L52) to get OpenGL texture handle from Diligent texture

Maybe there is an other way around, you create the context with Diligent and then make Qt use existing OpenGL context.

**Reply by AnloWho** [reply]
I also managed to get it to work in Qt. Currently the widget inherits from QOpenGLWidget and I use the CreateDeviceAndSwapChainGL() method provided by Diligent.

**Reply by TheMostDiligent** [reply]
Did you manage to make it work with other backends like Vulkan or Direct3D?

**Reply by AnloWho** [reply]
I will eventually try the Vulkan backend, that's where I want to end up. I need the tool to be cross-platform so I won't try Direct3D. 
Right now I'm trying to figure out how to get  Diligent to go for the discrete GPU, I've seen the comments in the code about using the GLAdapterSelector, so that's where I am right now. But I have good hopes about this.

**Reply by promerosenabre** [reply]
> I also managed to get it to work in Qt. Currently the widget inherits from QOpenGLWidget and I use the CreateDeviceAndSwapChainGL() method provided by Diligent.

Hi AnlOwho, I tried to run dirigent engine on Qt with QOpenGLWidget , but my knowledge is not enough.
Could you send me a basic example of how you achieved this? I'd really appreciate it. Thanks !!

**Reply by TheMostDiligent** [reply]
Yes, there is no official way to select an adapter on GL, but there are Nvidia and amd-specific workarounds, which seem to work

**Reply by AnloWho** [reply]
> > I also managed to get it to work in Qt. Currently the widget inherits from QOpenGLWidget and I use the CreateDeviceAndSwapChainGL() method provided by Diligent.
> 
> Hi AnlOwho, I tried to run dirigent engine on Qt with QOpenGLWidget , but my knowledge is not enough. Could you send me a basic example of how you achieved this? I'd really appreciate it. Thanks !!
I basically do this for my so called DiligentViewport class:
``class DiligentViewport : public QOpenGLWidget, protected QOpenGLFunctions``
`{`
`public:`
`    DiligentViewport(QWidget* parent = nullptr);`
`    ~DiligentViewport();`

`    virtual void paintGL() override;`
`    virtual void resizeGL(int width, int height) override;`
`    virtual void initializeGL() override;`

`private:`
`    bool initializeDiligentEngine();`
`    void createPipelineState();`
`    void createVertexBuffer();`
`    void createIndexBuffer();`

in the initializeGL() method I would run all the private methods in order of appearance. 
I'm still discovering, so I don't say this is the correct way, but I got it to work with the first triangle tutorial. It shows inside the MainWindowView (that inherits from QMainWindow). A crucial point was this cast to get the handle from the application windows down to the engine (in initializeDiligentEngine())
![image](https://github.com/user-attachments/assets/609965a4-e9ae-452e-8301-ad03240fb27b)

I don't do swapChain_->present() anywhere. I don't know if any of this I just mentioned is useful. Maybe it doesn't work for more complex scenarios. But if I can get it to work with the cube , why not the rest?

**Reply by promerosenabre** [reply]
Hi, if triangle tutorial is working, I guess won't be a problem to run all samples, Let me try it and I will inform you asap. Thank you very 
much !!

**Reply by AnloWho** [reply]
OK. I'm struggling with the Cube. I realized it is very important (on my system) to set the path to the folder where you have the shaders. And that path has to be absolute it seems. Use the method `CreateDefaultShaderSourceStreemFactory()` for your engine factory.

**Reply by TheMostDiligent** [reply]
Diligent does not do anything special with files. Shaders are just files, they need to be somewhere the system can find them. The paths may be relative or absolute.

**Comment by AnloWho** [comment]
Any idea why I get this error in the `render()` method when trying to setup the MapHelper?
```Diligent Engine: ERROR in Diligent::BufferGLImpl::MapRange() (BufferGLImpl.cpp, 374): glMapBufferRange() failed GL Error Code: 1281```
```Diligent Engine: ERROR: Debug assertion failed in Diligent::BufferGLImpl::MapRange(), file BufferGLImpl.cpp, line 374:```
I'm on my littel Qt test app where I before managed to draw the triangle from tutorial 01.
It's the line 253 in the Tutorial02_Cube.cpp.
```
    {
        // Map the buffer and write current world-view-projection matrix
        MapHelper<float4x4> CBConstants(m_pImmediateContext, m_VSConstants, MAP_WRITE, MAP_FLAG_DISCARD);
        *CBConstants = m_WorldViewProjMatrix;
    }
```

**Reply by TheMostDiligent** [reply]
Are there any other messages in the output?
Since annoyingly the error in OpenGL remains until the time when `glGetError` is called, this error may not be caused by the `glMapBufferRange`. It may be an error from some previous operation.

**Comment by AnloWho** [comment]
Sorry for bugging you, but abandoned the Opengl integration into Qt for now and have looked at using Vulkan instead. I got the clear color to show up in my Qt window, but I have these warning (in debug build) which suggest something is wrong:
_Diligent Engine: Warning: Render target view '0x13C197717A0' is not bound to the device context. ClearRenderTarget command is more efficient if render target view is bound to the device context. In OpenGL backend this is a requirement._
and  
_Diligent Engine: Warning: Depth-stencil view 'Default DSV of texture 'Main depth buffer'' is not bound to the device context. ClearDepthStencil command is more efficient when depth-stencil view is bound to the context. In OpenGL backend this is a requirement._

I know it can be difficult to answer since you don't have all the code, but I have used the handle from the QWindow to create a Win32NativeWindow before calling CreateSwapChainVk. 

Also, the ```immediateContext->Draw(drawAttrs)``` fails in the ```PrepareForDraw()``` with _No render pass is active while executing draw command_.  
Any pointer what I should do?

**Reply by AnloWho** [reply]
ok, I just put in:
```immediateContext_->SetRenderTargets(1, &pRTV, pDSV, RESOURCE_STATE_TRANSITION_MODE_TRANSITION);```
after clearing the RenderTarget and DepthStencil and now I got this :)
<img width="748" height="584" alt="image" src="https://github.com/user-attachments/assets/c5b57fb4-acfd-4552-915a-15d4c3a21390" />
I still have those Warnings in the Debug console as described above though...

**Reply by TheMostDiligent** [reply]
You need to set the render targets _before_ clearing them. This is what the error message is trying to say, may it needs to be more clear.

**Reply by AnloWho** [reply]
I don't know. I just missed that part looking at the Tutorials. My bad. I still get weird memory leak warning on closing the app. I need to understand in what order to tear down things here. Is there a common practice in Diligent on what to do on object destruction? I see Flush and Release being used.

**Reply by TheMostDiligent** [reply]
Calling Flush is reasonable.
On release, you should destroy all objects you created. If you use the smart pointer, they will destroy the objects automatically.
